### ISSUE TITLE
How do you get the camera intrinsic?

### ISSUE THREAD

**Issue by sta105** [geopavlakos/c2f-vol-demo#6]
Hi,

I found K is save in valid.mat, but I couldn't find it in human3.6m dataset. Could you tell me how can I get camera parameters for each video sequence? Thank you:)

Sicong

**Comment by nulledge** [comment]
Human3.6M provides some useful MATLAB scripts. The ```H36MCamera``` class is one of them. With this, you can get every camera intrinsic(F, C, K and P) by simply calling ```camera.f```, ```camera.c```, ```camera.p```, ```camera.k``` in MATLAB. The ```H36MCamera``` class is in ```{scripts root}/H36M/H36MCamera.m```.

**Comment by axhiao** [comment]
hi @nulledge do you know how to get the intrinsic and extrinsic TOF depth camera parameter? Thank you!

**Comment by hanabi7** [comment]
@axhiao Did you find the camera parameter of the top data? I understand the extrinsic parameters of the tof data is the same with rgb camera one, means they have the same rotation and translation matrix. But i can't obtain the intrinsic parameter of the tof data.

**Comment by axhiao** [comment]
> @axhiao Did you find the camera parameter of the top data? I understand the extrinsic parameters of the tof data is the same with rgb camera one, means they have the same rotation and translation matrix. But i can't obtain the intrinsic parameter of the tof data.

yes, tof camera is very close to one of rgb cameras. Neither do I get the intrinsic parameters. If you get the dataset officially, you could send them an email about the depth camera parameters. And if you get it somehow, please share a copy with me. I'd appreciate it!

**Comment by hanabi7** [comment]
@axhiao It seems that the focal length of the tof data can be roughly calculated based on the fact that the rgb camera and the tof data has the same range of picture if you get what i mean. Have you ever try that?

**Comment by axhiao** [comment]
@hanabi7  I tried cx, cy = width/2, height/2 and fx=fy=110. It seems to make sense visually. How about your choice?

**Comment by hanabi7** [comment]
basicly, the same. but the human length don't seem to make sense, and i couldn't ailgn it with the smpl parameters
| |
刘冠泽
|
|
***@***.***
|
签名由网易邮箱大师定制
On 3/12/2021 11:04，Minglei ***@***.***> wrote：

@hanabi7 I tried cx, cy = width/2, height/2 and fx=fy=110. It seems to make sense visually. How about your choice?

**Comment by hanabi7** [comment]
@axhiao The results seem to make sense, but did you find a way to align the tof point cloud with 3d poses or smpl model?

**Comment by hanabi7** [comment]
@axhiao  I think we could use 2d joint position for accurate focal_length prediction?

**Comment by axhiao** [comment]
@hanabi7 No. The point cloud is under the camera coordinate, but 3d keypoint is under world coordinate. I think you need a transformation matrix, including rotation and translation,  from camera coordinate to world coordinate. 

They seem not to provide 2d joint position on depth image.

**Comment by hanabi7** [comment]
@axhiao The transformation matrix is the same as the camera 55011271. It's mentioned in the metadata.xml i think? Let me know if you have any update.

**Comment by hanabi7** [comment]
@axhiao I find a way to transform the human3.6 2d joints position on rgb picture to 2d joints position on tof picture. I think that would help
![joints](https://user-images.githubusercontent.com/55522131/110918993-b90d1300-8356-11eb-8519-050c667b3c23.jpg)
Try focal_length = 248, fx = 110, fy=80

**Comment by axhiao** [comment]
@hanabi7 That's awesome! I think you mean focal_length=248, cx=110, cy=80? 
And did you also use the extrinsic parameter of  camera 55011271 to transform 3d coordinates to 2d coordinates on depth image?

**Comment by hanabi7** [comment]
@axhiao  Yes, but the focal length and cx, cy that i am getting is pure through calculation, it might be mistaken.

**Comment by axhiao** [comment]
@hanabi7 thank you! I think that's very close from the picture you posted above! I'll try those parameters!

**Comment by axhiao** [comment]
@hanabi7 Could you share the extrinsic parameter with me here? The result i'm getting seems not right.

**Comment by hanabi7** [comment]
@axhiao Rotation parameters: [[0.9216646531492915, 0.3879848687925067, -0.0014172943441045224], [0.07721054863099915, -0.18699239961454955, -0.979322405373477], [-0.38022729822475476, 0.9024974149959955, -0.20230080971229314]]
Translation parameters: [[-11.93434847209049], [449.41658936445646], [5541.113551868936]]
This is the extrinsic parameter that i am using, it belongs to sequence 11 and camera 55011271

**Comment by hanabi7** [comment]
@axhiao  https://github.com/karfly/human36m-camera-parameters This repository can be helpful

**Comment by axhiao** [comment]
@hanabi7 thank you so much! I'll try again.

**Comment by axhiao** [comment]
@hanabi7 I have tried with the parameters you provided. I think there should be a little differences between different subjects about the intrinsic parameters. Because some subjects' joints cannot fit well on depth image after projection from 3d to 2d. 

![image](https://user-images.githubusercontent.com/6879331/110993166-efda2d80-8344-11eb-8fef-b527705e8fff.png)

**Comment by hanabi7** [comment]
@axhiao Can you name the subject's that's working and the subject's that's not working ?
The picture that i posted above was the discussion scene from subject 11

**Comment by axhiao** [comment]
@hanabi7 for example, even for subject 11, this frame **S11_SittingDown 1_769** does not fit well either with your parameters. 

![image](https://user-images.githubusercontent.com/6879331/111016435-02bc2480-837c-11eb-9ea3-6d3037036e8d.png)

I manually adjust the intrinsic parameters for all subjects. The intrinsics work well for most of cases, but not well on all cases. You could use them as a reference.

```python
S1 = np.array([
        [248, 0,  100.0],
        [0, 248, 75],
        [0, 0, 1.0]
    ])

S5 = np.array([
        [248, 0, 102],
        [0, 248, 75],
        [0, 0, 1.0]
    ])

S6 = np.array([
        [248, 0, 103.5],
        [0, 248, 76],
        [0, 0, 1.0]
    ])

S7 = np.array([
        [248, 0,  102.5],
        [0, 248, 75],
        [0, 0, 1.0]
    ])

S8 = np.array([
        [248, 0,  105.8],
        [0, 248, 79],
        [0, 0, 1.0]
    ])

S9 = np.array([
        [248, 0,  107.5],
        [0, 248, 78.5],
        [0, 0, 1.0]
    ])

S11 = np.array([
        [248, 0,  108],
        [0, 248, 80],
        [0, 0, 1.0]
    ])
```

**Comment by hanabi7** [comment]
@axhiao  That's very helpful

**Comment by hanabi7** [comment]
@axhiao I wonder how did you transform the cam coordinates point clouds to world coordinates point cloud? 
I tried world_coordinates = np.dot(cam_coordinates - extrinsic_translation, np.linalg.inv(extrinsic_rotation)) but it seems not working?

**Comment by axhiao** [comment]
@hanabi7  

image is the depth map

```python
def depthmap2points(image, fx, fy):
    h, w = image.shape
    x, y = np.meshgrid(np.arange(w) + 1, np.arange(h) + 1)
    points = np.zeros((h, w, 3), dtype=np.float32)
    points[:,:,0], points[:,:,1], points[:,:,2] = pixel2world(x, y, image, w, h, fx, fy)
    return points


def pixel2world(x, y, z, img_width, img_height, fx, fy):
    # w_x = (x - img_width / 2) * z / fx
    # w_y = (img_height / 2 - y) * z / fy
    w_x = (x - 104.1857) * z / fx
    w_y = (76.9285 - y) * z / fy
    w_z = z
    return w_x, w_y, w_z
```

**Comment by hanabi7** [comment]
@axhiao Do you know how to align the given 3d smpl model with the generated point cloud? Even if I transform the coordinates in the world coordinate system, point cloud and the smpl model still has a giant gap in rotation. 
![snapshot00](https://user-images.githubusercontent.com/55522131/111063037-82b2cd80-84e7-11eb-8603-daf6cd538b76.png)
The generated point cloud is meshed for better visualization results

**Comment by axhiao** [comment]
@hanabi7 sorry, I know little about SMPL.

**Comment by patriciamdr** [comment]
@axhiao @hanabi7 May I asked how you preprocessed the TOF data? I wasn't able to find any open source code for it and following the MATLAB code of Human3.6M data only results in black and white images for me. Thank you!

**Comment by patriciagschossmann** [comment]
@hanabi7 @axhiao May I asked you how you obtained the c_x and c_y values? Did you use a formula or did you manually tested different values?

**Comment by axhiao** [comment]
Hi @patriciamdr If you have the right to access the dataset H3.6M, you could ask them if they provide the camera parameters for the TOF data. But as far as I know, they probably do not have it.

**Comment by Dipankar1997161** [comment]
@axhiao 
@hanabi7 
How do get the human3.6m extrinsic camera parameters? I checked the repo mentioned above, however, it doesn't look quite accurate to me.

Did anyone else validate? Or does anyone know how the intri and extr camera parameters of h36m dataset? 

Would be really useful for my research. THank you so much

**Comment by nulledge** [comment]
@Dipankar1997161 
I think that repository provides accurate parameters. But if you want to get parameters manually, see ```H36M/H36MCamera.m```. All parameters are stored in ```metadata.xml``` and ```H36M/H36MCamera.m``` extracts camera parameters(f, c, k, p, R, T) from it.

If you are not familiar with MATLAB, [my Python script(link)](https://gist.github.com/nulledge/07c8337d16762793f087473a2ef0cad9) may be helpful. Install MATLAB engine library and just call ```get_intrinsics``` function to get f, c, k, p. Also you can edit it to get R and T.

Note that Human3.6M provides only parameters of RGB, but not TOF.  It suggests to use parameters of RGB as approximated parameters of TOF. Of course it's not accurate.
